### ISSUE TITLE
Simulator Flickering Issue with Latest Flutter and iOS Versions

### ISSUE THREAD

**Issue by sinnoorc** [flutter/flutter#148660]
### Steps to reproduce

After updating Flutter to version 3.22 and iOS to version 17.5, I've encountered a bug in the iOS simulator. The simulator screen flickers whenever an animation occurs, such as when a loading indicator is displayed. This issue did not exist before the updates. 

1. **Update Flutter and iOS:**
   - Ensure your Flutter is updated to version 3.22.
   - Ensure your iOS is updated to version 17.5.

2. **Open a Flutter Project:**

3. **Add an Animation:**

4. **Run the Project in the iOS Simulator:**
   - Open the iOS simulator.
   - Run the Flutter project using the command `flutter run`.

5. **Observe the Issue:**
   - Observe the simulator for any screen flickering when the animation (e.g., the loading indicator) is running.

## Environment

- **Flutter version:** 3.22
- **iOS version:** 17.5
- **Simulator device:** [iPhone 15 pro]

### Expected results

The iOS simulator should run smoothly without any screen flickering during animations. For example, when a loading indicator is displayed, the animation should appear seamlessly without any visual disturbances. The application should function as intended, providing a stable and clear user experience.

### Actual results

After updating Flutter to version 3.22 and iOS to version 17.5, the iOS simulator exhibits screen flickering whenever an animation occurs. For example, when a loading indicator is displayed, the screen flickers continuously, disrupting the animation and overall user experience. This issue was not present before the updates.

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

     void main() {
       runApp(MyApp());
     }

     class MyApp extends StatelessWidget {
       @override
       Widget build(BuildContext context) {
         return MaterialApp(
           home: Scaffold(
             appBar: AppBar(title: Text('Flicker Test')),
             body: Center(
               child: CircularProgressIndicator(),
             ),
           ),
         );
       }
     }
```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

https://github.com/flutter/flutter/assets/62532648/6da28b92-c64a-4580-a737-c22f552cefde

</details>

### Logs

<details open><summary>Logs</summary>

[https://pastebin.com/BjRKbn8J](url)

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
flutter doctor -v
[✓] Flutter (Channel stable, 3.22.0, on macOS 14.5 23F79 darwin-x64, locale en-IN)
    • Flutter version 3.22.0 on channel stable at /Users/macsho/development/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 5dcb86f68f (11 days ago), 2024-05-09 07:39:20 -0500
    • Engine revision f6344b75dc
    • Dart version 3.4.0
    • DevTools version 2.34.3

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/macsho/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = /Users/macsho/Library/Android/sdk
    • ANDROID_SDK_ROOT = /Users/macsho/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11572160)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.4)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15F31d
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11572160)

[✓] VS Code (version 1.89.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.89.20240501

[✓] Connected device (4 available)
    • Sinnoor’s iPhone (mobile) • 00008110-0016391E2E29401E            • ios            • iOS 17.5 21F79
    • iPhone 15 Pro (mobile)    • 057185C5-1CED-4300-9FDE-24444626C307 • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-17-5 (simulator)
    • macOS (desktop)           • macos                                • darwin-x64     • macOS 14.5 23F79 darwin-x64
    • Chrome (web)              • chrome                               • web-javascript • Google Chrome 124.0.6367.208

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

**Comment by Khushnu** [comment]
same issue am encountering

**Comment by darshankawar** [comment]
Thanks for the report @sinnoorc 
It seems to be OS version specific, as I wasn't able to replicate this on iOS 15.
Can you check if this only occurs using impeller or without it as well ? You can disable it as below in `info.plist` and re-run your scenario to confirm.

```

<key>FLTEnableImpeller</key>
<false/>

```

**Comment by sinnoorc** [comment]
@darshankawar  
I disabled Impeller as per your instructions by adding <key>FLTEnableImpeller</key><false/> to the info.plist file. After making this change, the issue no longer occurs and the simulator runs smoothly without any screen flickering during animations.

Thank you for the assistance!

**Comment by sinnoorc** [comment]
@darshankawar 
after adding this am getting this warning 

```console
[IMPORTANT:flutter/shell/common/shell.cc(454)] [Action Required] The application opted out of Impeller by either using the --no-enable-impeller flag or FLTEnableImpeller=false plist flag. This option is going to go away in an upcoming Flutter release. Remove the explicit opt-out. If you need to opt-out, report a bug describing the issue.
[IMPORTANT:flutter/shell/platform/darwin/graphics/FlutterDarwinContextMetalSkia.mm(66)] Using the Skia rendering backend (Metal).
```

**Comment by huycozy** [comment]
I can reproduce this issue on iOS 17.4 emulator (haven't checked this on a physical device as I don't have one). I narrow this down a bit and see the issue seems to appear with AnimatedBuilder which is used internally inside CircularProgressIndicator.

The issue occurs on the latest stable 3.22.0 and master channel whilst it works normally on the previous 3.19.6.

| with CircularProgressIndicator | with AnimatedBuilder |
| --------------- | --------------- |
<video src="https://github.com/flutter/flutter/assets/104349824/17b9201f-8399-4b00-8eee-671864706feb"/> | <video src="https://github.com/flutter/flutter/assets/104349824/1386fc2c-38d1-4d60-a990-e4e983c1a68d"/>

<details>
<summary>flutter doctor -v (stable and master)</summary>

```bash
[✓] Flutter (Channel stable, 3.22.0, on macOS 14.1 23B74 darwin-x64, locale en-VN)
    • Flutter version 3.22.0 on channel stable at /Users/huynq/Documents/GitHub/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 5dcb86f68f (11 days ago), 2024-05-09 07:39:20 -0500
    • Engine revision f6344b75dc
    • Dart version 3.4.0
    • DevTools version 2.34.3

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.3)
    • Xcode at /Applications/Xcode15.3.app/Contents/Developer
    • Build 15E204a
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • android-studio-dir = /Applications/Android Studio.app/
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)

[✓] VS Code (version 1.89.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.88.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 14.1 23B74 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 124.0.6367.208

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

```bash
[!] Flutter (Channel master, 3.22.0-39.0.pre.1, on macOS 14.1 23B74 darwin-x64, locale en-VN)
    • Flutter version 3.22.0-39.0.pre.1 on channel master at /Users/huynq/Documents/GitHub/flutter_master
    ! Warning: `flutter` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 165e535474 (3 hours ago), 2024-05-20 16:52:15 -0700
    • Engine revision c2ef01f6f1
    • Dart version 3.5.0 (build 3.5.0-172.0.dev)
    • DevTools version 2.36.0-dev.10
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.3)
    • Xcode at /Applications/Xcode15.3.app/Contents/Developer
    • Build 15E204a
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • android-studio-dir = /Applications/Android Studio.app/
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)

[✓] VS Code (version 1.89.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.88.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 14.1 23B74 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 125.0.6422.61

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```
</details>

**Comment by mauricev** [comment]
When I put up a showdialog and type into a text field, i see all manner of screen artifacts including the widgets in the screen behind disappearing. When i disabled impeller per the suggestion above, these problems cease. I am using ipad simulator with  17.2.

**Comment by gaaclarke** [comment]
I was able to verify that the flickering doesn't happen on device (iPhone SE 2022 - iOS 17.4.1) at 8eba63f29b8b26931496807cb4fc2fed8e4e3258

I was also able to verify that the flickering doesn't happen on iPhone 15 Pro on iOS 17.4 (Xcode 15) at the same hash.

I was able to test on iPhone 15 Pro on iOS 17.5 (Xcode 15.4) at the same hash and it also worked:

<img src="https://github.com/flutter/flutter/assets/30870216/ae810424-1eb0-429c-949c-b73857b25520" width=150/>

**Comment by gaaclarke** [comment]
I tested with 3.22 stable as well and I don't see the flickering on the iPhone 15 Pro iOS 17.5 simulator.

**Comment by navaronbracke** [comment]
I do see it on stable 3.22.0, when running with Impeller only.

I am using XCode `15F31d` and the `Iphone 15 iOS 17.5` Simulator.

**Comment by gaaclarke** [comment]
Okay, here's what we know:
1) The glitch doesn't happen on device
1) The glitch only happens on some macs

Here's the next question:  For the people that are seeing it, what kind of Mac are you using.  I have an M1 Max.

This is sounding like an Apple bug unfortunately.  Someone that can reproduce it should file an apple bug and try downgrading xcode to 15.3 for now.  I'm unable to reproduce the problem so I can't test if that is a valid workaround.  If someone does verify that, could they post that info here please for others?

**Comment by sinnoorc** [comment]
I am running on a Mac with an Intel chip.

**Comment by gaaclarke** [comment]
> I am running on a Mac with an Intel chip.

That's what I was afraid of.  I don't know how successful you will be at getting Apple to address this since it has stopped selling intel macs a while ago.  I'm going to close this since there is nothing we can do from flutter's side to address this.

The workaround is to test on device if you have an intel mac.

**Comment by mauricev** [comment]
Before 3.22, Flutter was working fine. And we know it's triggered only in the Impeller engine. At the same time, Apple is still producing Sonoma builds for Intel Macs.

**Comment by gaaclarke** [comment]
Reopening this so the bot doesn't lock this in case more information comes in.

**Comment by jason-simmons** [comment]
Reproduced this on an Intel Mac Mini.  Bisected it to https://github.com/flutter/engine/commit/431d64dbb4b8c3c56e37f5a8cf8f72aa789dc97c

@jonahwilliams

**Comment by jonahwilliams** [comment]
So this is likely a problem with the storage mode + flush implementation for simulators. I'm not sure what kind of buffer storage mode we're getting on the intel simulators. I don't know if FML_OS_IOS is enabled for simuators as well, but if not then the DeviceBuffer::Flush implementation might be broken:

https://github.com/flutter/engine/blob/main/impeller/renderer/backend/metal/device_buffer_mtl.mm#L57-L65

**Comment by gaaclarke** [comment]
> I don't know if FML_OS_IOS is enabled for simuators as well

`FML_OS_IOS` will be defined as 1 for any target of iOS (iphone, ipad) and that includes builds that run on simulators or devices.

**Comment by jonahwilliams** [comment]
Okay then that code is wrong, we need to execute the flush for sim targets

**Comment by jason-simmons** [comment]
In the scenario where this happens it looks like `HostBuffer` is reusing device buffers before Metal is done consuming them.

Tried adding some logging to `HostBuffer::Reset` and an `addCompletedHandler` callback on the `MTLCommandBuffer`.  The rate of `HostBuffer::Reset` calls exceeds the completion handler calls.

Each `HostBuffer::Reset` advances the `HostBuffer`'s frame index.  The frame index rotates among a pool of `kHostBufferArenaSize` (currently 3) sets of buffers that are reused.  So if `HostBuffer::Reset` is not evenly matched with frame completion then eventually the frame index will wrap around to a buffer that is still in use by Metal.

I was able to get this to run smoothly on the Intel Mac Mini iOS simulator with this patch:
* add an `fml::Semaphore` initialized with count=3 to `surface_mtl.mm`
* call `Wait` on the semaphore before the `CAMetalLayer.nextDrawable` call in `SurfaceMTL::GetMetalDrawableAndValidate`
* in `SurfaceMTL::Present` call `command_buffer.addCompletedHandler` with a callback that signals the semaphore

**Comment by emrszr** [comment]
I faced the issue on an an Intel chip Mac.

**Comment by jacekendrick** [comment]
🙋 Another Intel mac user with the issue

**Comment by jonahwilliams** [comment]
Cherry pick pending at https://github.com/flutter/flutter/issues/149700

**Comment by Khushnu** [comment]
another mac intel core i5 user here facing same problem after upgrading flutter sdk also xcode 15.4 and emulator to 17.5

**Comment by IbrokhimMirzaev** [comment]
how to fix this issue now ??

**Comment by martin-braun** [comment]
![2024-06-13 18 38 04](https://github.com/flutter/flutter/assets/24259245/408d9253-f438-4470-b69c-ed7753cacee7)

This is due to animations in the horizontal `ListView` with `PageScrollPhysics`.

I tested all `3.22.X` versions. `3.19.6` did not produce those issues.

**Comment by Asharulislam** [comment]
I faced the issue on an an Intel chip Mac.

**Comment by DavydNemyrych** [comment]
Same problem on my iOS Simulator 17.5 (Xcode 15.4).
 I have a MacBook Pro 2.6 GHz 6-Core Intel Core i7.
 After I upgraded Flutter to v3.22.2, the glitches began, and in some cases, the Mac became unresponsive and restarted on its own. So the issue is quite severe.

**Comment by AdamBridges** [comment]
> ![2024-06-13 18 38 04](https://github.com/flutter/flutter/assets/24259245/408d9253-f438-4470-b69c-ed7753cacee7)
> 
> This is due to animations in the horizontal `ListView` with `PageScrollPhysics`.
> 
> I tested all `3.22.X` versions. `3.19.6` did not produce those issues.
> 
> @IbrokhimMirzaev https://github.com/flutter/flutter/issues/148660#issuecomment-2120389464 next time read on. This fixes the problem for now.

The issue still occurred for me on `3.19.6`. My setup was on Sonoma iOS v15.5; however, I'm using opencore with a late 2014 mac mini model, so that could also be a factor.

**Comment by mths0x5f** [comment]
Another Intel Mac user facing this issue.

**Comment by chahedmoetaz** [comment]
Certainly! It seems you encountered an issue with the iOS simulator after upgrading to Flutter version 3.22. Let's address this step by step:

1. **The Issue:**
After updating Flutter to version 3.22 and iOS to version 17.5, you noticed screen flickering in the iOS simulator during animations (e.g., loading indicators). This issue did not exist before the updates¹.

2. **Solution:**
To mitigate this issue, follow these steps:
- Open your Flutter project in Xcode using the `.xcworkspace` file (instead of `.xcodeproj`). The `.xcworkspace` contains information about pods and should help resolve the flickering problem².
- Additionally, consider disabling the impeller by adding the following key-value pair to your `ios/Runner/Info.plist` file:
```xml
<key>FLTEnableImpeller</key>
<false/>
```
This should help stabilize the simulator behavior³.

3. **Share with Your Team:**
Feel free to share this solution with your team. Let them know about the `.xcworkspace` preference and the impeller configuration. Hopefully, this will help others who encounter similar issues.

Remember, it's essential to keep your development environment up-to-date and share knowledge within your team to improve productivity and avoid common pitfalls. Happy coding! 🚀📱💻¹²³

https://github.com/flutter/flutter/assets/43938322/958b4382-4be6-4d85-b8fb-f5b96a3905c0

**Comment by jmagman** [comment]
I strongly advise not turning off Impeller on your production app just for the sake of not seeing glitching in debug mode on a simulator...

**Comment by martin-braun** [comment]
@chahedmoetaz I agree with @jmagman. Afaik, the Impeller was introduced to iOS to combat weak render performances on iOS. Disabling it and thus going back to the old renderer worked for us, but it essentially also uses an obsolete technology in production. We want to move away from this solution asap.

Is it maybe possible to tell the iOS project to disable the Impeller only on the Simulator at least?

**Comment by chahedmoetaz** [comment]
Hello @martin-braun @jmagman,

thank you for interacting with the post, but this solution will fix the debug bug also it works on production because the real Ios device had large storage and the software works better than the simulator cuz It limited storage that's why many developers got this bug on big applications 

thank you.

**Comment by jonahwilliams** [comment]
The issues has been fixed on master, and should be fixed in the upcoming beta and stable. We may hot fix this in 3.22, but there are no specific plans for this yet.

**Comment by jmagman** [comment]
The fix is available on Flutter 3.22.3.  Thank you!
